### PR TITLE
Modifications to zmat

### PR DESCRIPTION
**Pull request by alongd**
Consider higher constraints when determining atom order to allow more flexibility in defining zmats, and other robustness fixes.

For the Reviewer(s):
"Constraints" are inputs we give when creating a zmat, and the created zmat will be defined in such a way that the requested constraining parameters will be part of it.
For example, if one wants to modify an angle in a molecule and all the groups connected to the last atom, we will pass a respective `A_group` constraint (A for angle, we can also use R for distance or D for dihedral) as the key in the `constraints` dict, and the value will be a list of the respective atom indices (in this case, three atoms).

This PR fixes how we treat constraints. It specifically fixes the function that determines the first reference atom `determine_r_atoms()`. It allows A and D constraint consideration in addition to R constraints, which means that if one asked for an A or D constraint, the same requested atoms will also appear in the R value to be consistent in the definition of the zmat.

### CODE REVIEW

**Comment by alongd** [comment]
@oscarwumit, @xiaoruiDong, @mefuller,
Could one of you review this PR? It's relatively short, merging it will assist in advancing the `job` branch. Thanks!!

File: arc/species/zmat.py [hunk 1]
```diff
@@ -200,9 +215,14 @@ def determine_r_atoms(zmat: Dict[str, Union[dict, tuple]],
...
-        # 1. Always use the constrains if given.
-        r_atoms = [n] + [key_by_val(zmat['map'], atom) for atom in r_constraint[1:]]
+    elif any(constraint is not None for constraint in [r_constraint, a_constraint, d_constraint]):
+        # 1. Always use the constraint if given.
+        if r_constraint is not None:
+            r_atoms = [n] + [key_by_val(zmat['map'], r_constraint[1])]
+        elif a_constraint is not None:
+            r_atoms = [n] + [key_by_val(zmat['map'], a_constraint[1])]
+        elif d_constraint is not None:
+            r_atoms = [n] + [key_by_val(zmat['map'], d_constraint[1])]
```
[the remark this answers is not in the document]
Comment (alongd) [reply arc/species/zmat.py:225]: I think it's OK - L223 uses `a_constraint`, and L225 uses `d_constraint`. They'll be used by order using the `elif` condition

File: arc/species/zmat.py [hunk 2]
```diff
@@ -333,9 +356,13 @@ def determine_a_atoms(zmat: Dict[str, Union[dict, tuple]],
         # This is the 1st or 2nd atom added to the zmat, there's no angle definition here.
         a_atoms = None
     elif a_constraint is not None:
+        # always use the constraint if given
         if a_constraint_type not in ['A_atom', 'A_group', None]:
             raise ZMatError(f'Got an invalid A constraint type "{a_constraint_type}" for {a_constraint}')
         a_atoms = [n] + [key_by_val(zmat['map'], atom) for atom in a_constraint[1:]]
+    elif d_constraint is not None:
+        # consider d_constraint for consistency if a_constraint is None
+        a_atoms = [n] + [key_by_val(zmat['map'], atom) for atom in d_constraint[1:3]]
```
[the remark this answers is not in the document]
Comment (alongd) [reply arc/species/zmat.py:365]: Sure. Starting from the 4th atom in a zmat, we have to give four atom indices to describe it: the "self" index (trivial), an index of an atom from which we measure distance (called r atom), an index to another atom for defining an angle (the self -- r atom -- a atom angle), and an index for another atom for defining a dihedral angle (the self -- r atom -- a atom -- d atom dihedral angle).

If we have constraints in setting up the zmat, we must consider them.

In `determine_r_atoms()` we need to determine the r atom. If we get a constraint request of some R distance from some atom, we should use that atom as r atom (trivial). But if we get a more complex constraint of a atom or d atom, then we should be able to extract r atom from there (as shown in parenthesis in the beginning of this comment). These are essentially the additions I made. They look similar, but they differ in the nuances: whether we're considering `a_constraint` or `d_constraint`.

Let me know if this sounds reasonable, I'm happy to elaborate more.

**Comment by alongd** [comment]
@mefuller, I rebased, please take a final look. Let me know of course if you have additional comments. Thanks!

**Comment by mefuller** [comment]
I think everything is fine now / feel free to merge

**Review by mefuller — APPROVED** [review]

**Pull request merged**